### PR TITLE
[python/ci] Add CI job to test building wheel from sdist

### PR DESCRIPTION
**Pull request by jdblischak**
**Issue and/or context:**

TileDB-SOMA-Py is built and distributed in many different ways (from source, PyPI wheels, conda binaries). We want to identify any packaging-related problems immediately in the PR that generates them instead of in a nightly build or worse after the release has been made.

**Changes:**

This PR includes 2 main changes:

* The first is rebranding my existing CI pipeline to `python-ci-packaging.yml`. The original name was from my initial PR to test the shared object copying behavior (#1937, #2220). However, I had already previously expanded the pipeline to test the pure setuptools installation setup (which we use in tiledbsoma-feedstock) (#2221). With the addition of a 3rd job (below), I wanted to update the CI to better match its purpose

* The 1.9.0 release was delayed because the PyPI wheels could not be built (#2337, #2339). While we build wheels in many of the existing CI pipelines, none build the wheel from a source tarball (sdist) as is done in [`python-packaging.yml`](https://github.com/single-cell-data/TileDB-SOMA/blob/36d5b6e7e9657ba9d41e9f71c0b19d0468dc1f79/.github/workflows/python-packaging.yml#L31), so this error wasn't caught until post release. With this new CI job, we will immediately detect if a crucial file is absent from the source tarball 

**Notes for Reviewer:**

See the wiki page [Debugging wheel‐build issues](https://github.com/single-cell-data/TileDB-SOMA/wiki/Debugging-wheel%E2%80%90build-issues) for more details

### CODE REVIEW

**Comment by johnkerl** [comment]
@jdblischak is this partially duplicating https://github.com/single-cell-data/TileDB-SOMA/blob/main/.github/workflows/python-packaging.yml ?

**Comment by jdblischak** [comment]
> is this partially duplicating https://github.com/single-cell-data/TileDB-SOMA/blob/main/.github/workflows/python-packaging.yml?

I wouldn't describe it as duplication. I would describe it more as preparation for the official building of the PyPI wheels.

Differences:

* Uses transparent python commands instead of the opaque action `pypa/cibuildwheel`, and therefore easier to debug
* Only builds two wheels total (python 3.11 on Ubuntu and macOS), so it is much faster. It's not creating the wheels for distribution, but instead just providing a quick check that all the required files are distributed in the source tarball

In summary, the goal is to identify any potential problems with building the PyPI wheels when a PR is submitted, and not at release time.

**Comment by johnkerl** [comment]
@jdblischak can we please get this verbatim into the affected YAML file as a block comment?

> Differences:
> 
> Uses transparent python commands instead of the opaque action pypa/cibuildwheel, and therefore easier to debug
Only builds two wheels total (python 3.11 on Ubuntu and macOS), so it is much faster. It's not creating the wheels for distribution, but instead just providing a quick check that all the required files are distributed in the source tarball
In summary, the goal is to identify any potential problems with building the PyPI wheels when a PR is submitted, and not at release time.

**Review by johnkerl — COMMENTED** [review]
@jdblischak SGTM as long as CI resource consumption on every PR doesn't become excessive ... main issue being macos runners: choices are 5, 5, 5, or enterprise with no "step up" to 10 or 20 say ...

**Review by johnkerl — APPROVED** [review]
@jdblischak SGTM as long as CI resource consumption on every PR doesn't become excessive ... main issue being macos runners: choices are 5, 5, 5, or enterprise with no "step up" to 10 or 20 say ...

**Comment by jdblischak** [comment]
> as long as CI resource consumption on every PR doesn't become excessive

This pipeline does spawn a lot of jobs. Two reasons why I think it avoids being excessive:

1. Aggressive use of path-based filters. This pipeline is only triggered if a file that could affect the Python installation process is affected (eg `setup.py`, `pyproject.toml`, etc). A PR that edits the C++ code, R code, or even the non-installation Python files will never trigger these jobs
2. The macOS jobs all finish in under 10 minutes. So even if it hogs our parallel macOS jobs, it will be for a brief period of time

If it does start to become excessive, the first macOS job I would drop is the one added in this PR, `sdist`. It is mainly to catch files missing from the source tarball, and that behavior should be consistent across Linux and macOS.

**Comment by johnkerl** [comment]
Cool and thanks @jdblischak ! I indeed just looked for `paths:` in your affected YAML and we have protection there against over-running so my concern was premature (I should have checked there first). But thanks for looking, and thanks for the prioritization if/when "something has to go" someday -- really appreciate it! :)
